### PR TITLE
feat(hooks): surface class-scope divergence from the diff, not from a round counter

### PR DESCRIPTION
**Pull request by WingedGuardian**
## Why

The commit gate already refuses a third consecutive defect-bearing review round with
*"you are fixing the INSTANCE the reviewer named, not the CLASS"*. That check works —
but it reads **no signal from the diff at all**. It triggers on a counter of rounds, so
it cannot speak until two rounds have already been spent, and it cannot see that a
single edit left its siblings behind.

These two scanners give it a content signal available on the very first edit.

## What they detect

**Orphaned literals** — a string literal is changed in one file while the identical
value survives elsewhere. AST-based rather than regex over diff text, and that choice is
load-bearing: a regex prototype excluded backslashes, so it could not see *any* prompt
string (they all end in `\n`) and cheerfully reported "no findings" while looking at
nothing. Only replaying a known-positive told the two apart.

**Changed provenance** — when an edit changes what a variable is assigned *from*, every
later use now reads something different. The scan enumerates those uses and reports the
ones the edit did not touch. It compares callee *identity* rather than the unparsed
expression, so rewrapping a receiver around the same call is not a finding.

## Wiring

A `PostToolUse` advisory on `Write|Edit` for latency — saying this while the edit is
fresh rather than once all the work is done — plus a separate `PreToolUse` advisory on
commit commands for a consolidated view. Both always exit 0. Neither is
`run_guard`-wrapped: that wrapper fails **closed** and is reserved for
irreversible-action guards, and an advisory that blocked an edit over its own parse
error would be worse than the problem it reports.

The commit-time check is a **separate hook**, not lines inside
`review_enforcement_commit.py`. An earlier revision put it there, and adversarial review
measured **5.98s** of `git grep` fan-out on a 20-file changeset — against that gate's
single 10s timeout, shared with its branch, marker, substantiality and escalation logic.
A timed-out `PreToolUse` hook does not deny, so an advisory could have starved a
mandatory gate into failing open on exactly the large commits where it matters most.
Separate hooks get separate budgets; the gate file is untouched. The scanner also
carries its own wall-clock budget, because cost is files × literals × repo size and
per-item caps do not bound it.

## Measured

Over 151 real file-edits from recent history:

| scan | fire rate | notes |
|---|---|---|
| orphaned literals | **0/151 (0.0%)** | after excluding identifier-shaped literals; the one hit before that filter was `run_in_background`, duplicated across a repo as a matter of course |
| changed provenance | **5/151 (3.3%)** | on inspection all five are genuine "this variable means something different now" signals — a fire rate, not an error rate |

Latency: worst-case commit path **5.98s → 0.53s** after the budget; edit path **0.71s**
on the two largest files in the repo, **0.17s** at the literal cap.

Both scanners were replayed against the two real review defects that motivated them and
name the exact sibling file and the exact unrevisited lines. Those replays are permanent
tests, and each was observed to fail when the scanner was neutered. The acceptance
replays were re-run at the tightest per-file budget actually used — a budget that
improved the latency number by blinding the scan would otherwise look identical.

## Also

Records the methodology in the `genesis-development` skill: an **acceptance bar** (replay
the real defect) plus a **measured rate** (`k/N` on a real corpus, tightened and
re-measured with the acceptance replay re-run), and the rule that a review's findings are
a **sample**, not a to-do list — name the class, enumerate the population, fix the
population.

## Review note

Cross-model review was attempted first per the standing rule; Codex is quota-exhausted
until 2026-09-10. The adversarial pass on this branch is therefore Claude-only, with the
blind-spot overlap that implies. It found one blocker (the gate-timeout issue above) and
three lesser defects, all fixed: sentinel files were landing in the watchdog-policed temp
directory, `UnicodeDecodeError` is not an `OSError` so one bad file aborted the whole
scan, and `git grep` needed `-e` so a literal starting with `-` was not silently skipped.

## Summary by CodeRabbit

* **New Features**
  * Added a class-scope scan to identify duplicated code changes that may have been missed.
  * Supports worktree and staged content, nested scopes, path filtering, scan limits, and coverage reporting.
  * Detects unchanged sibling behavior and variable uses that may require additional updates.
  * Clearly distinguishes clean results, skipped items, unreadable files, and scan errors.

* **Documentation**
  * Updated review guidance to require checking sibling instances before marking findings fixed.

* **Tests**
  * Added comprehensive coverage for scanning, edge cases, limits, and reporting.

### CODE REVIEW

**Comment by WingedGuardian** [comment]
## Automated first-pass review

**Structural review:** no ERROR-tier pattern violations. A few WARNING/NOTE items below.
**PII/Secrets/Wording scan:** CLEAN — no hard blocks, no soft warnings, no problematic language in the diff, title, or body.

---

### Structural — WARNING

1. **Dead code: `_tracked_python_files` is defined but never called.** `scripts/hooks/class_scope_scan.py:96` defines `_tracked_python_files(repo_root) -> list[Path]`, but nothing in the diff invokes it — `find_orphaned_literals` uses `_grep_candidates` (a direct `git grep`) instead of iterating this list. Looks like a leftover from an earlier implementation shape. Either wire it in or remove it.

2. **No test coverage for the hook entrypoints themselves.** `tests/test_hooks/test_class_scope_scan.py` thoroughly covers `class_scope_scan.py`'s scanning logic (including the acceptance-bar replays), and separately unit-tests `class_scope_advisory._SENTINEL_DIR` and `class_scope_commit_advisory._targets_a_commit`. But `main()` in both `scripts/hooks/class_scope_advisory.py` and `scripts/hooks/class_scope_commit_advisory.py` — the payload parsing, session-scoped sentinel dedup, and stderr output formatting — has no test exercising it end-to-end (e.g. via `hook_input.read_payload()` with a constructed stdin payload).

### Structural — NOTE

3. **`git add -A` appears repeatedly in test fixtures** (`tests/test_hooks/test_class_scope_scan.py`, e.g. the `repo` fixture and several tests). This matches the flagged `git add -A` pattern literally, but every occurrence operates on an ephemeral `tmp_path` git repo created solely for the test — not the real working tree — so there's no actual risk of staging unintended files. Flagging for visibility only; not blocking.

### Verified as correct (per PR description's own claims)

- `~/tmp/genesis-class-scope` sentinel location (not `/tmp`, not the CC-managed `cc-tmp`) — confirmed correct and tested (`TestSentinelLocation`).
- The commit-time advisory is wired as a genuinely separate `PreToolUse`/`Bash` hook entry in `.claude/settings.json` with its own `timeout: 10`, sitting after (not inside) `review_enforcement_commit.py` — confirmed `review_enforcement_commit.py` is untouched by this diff.
- Both new hooks are stdlib-only (no third-party imports), consistent with the hook-scripts-must-be-stdlib convention.
- No bare `except:`, no mutable default args, no raw SQL, no `nohup`, no `/tmp/` usage, no force-push/reset --hard, no `git add .`/`-A` outside test fixtures.

---


---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by WingedGuardian** [comment]
## Cross-model review round (Kimi Code K3)

Codex is quota-exhausted until 2026-09-10, so this branch went through **Kimi Code K3** (`kimi-code/k3`, high effort) with explicit per-use approval. It found **one blocker and five majors that the Claude-side review missed entirely** — which is the case for the cross-model rule rather than a formality.

| severity | finding | status |
|---|---|---|
| BLOCKER | the wall-clock budget was checked only *between* literals, never inside the survivor loop where the `ast.parse` work happens; each `git grep` also carried a fixed 5s timeout against a 1.5s budget. Measured **1.15s at `budget_seconds=0.001`** | fixed — deadline per candidate, subprocess gets `min(remaining, 5s)` |
| MAJOR | the escape set was a **blocklist**; every escape it omitted (`\xNN`, `\uXXXX`, octal) put a character in the grep core that never appears in raw source → silent false negative | fixed — now an allowlist of verbatim characters |
| MAJOR | functions keyed by **bare name**, so two classes with the same method name collided | fixed — qualified names |
| MAJOR | variable provenance was **last-assignment-wins**, so a later reassignment masked the change | fixed — a set of callees per variable |
| MAJOR | the commit advisory listed staged files but read the **worktree** | fixed — reads the index blob |
| MAJOR | an embedded NUL raised `ValueError`, which is neither `SubprocessError` nor `OSError` | fixed |

Two of those are the same errors this PR exists to catch: a **blocklist that silently skips what it forgets**, and a **budget that isn't a budget**. The bare-name collision is sharper still — the file that motivated this scanner has the same method name in two builder classes, so the detector would have missed its own motivating defect.

**Class sweep, not instance patching.** The reviewer named one `ValueError` site; three more existed and are fixed with it. Time-bounding was audited across every path — the orphan scan needs a deadline because its cost scales with repo size, while the provenance scan is two parses and a difflib pass over one file (measured 0.28s worst case on the two largest files here), so that is documented with the measurement rather than given untestable ceremony.

**Test gaps closed.** Nothing previously executed either hook's `main()`, so deleting the dedup or the whole provenance call left the suite green. Both hooks now run as real subprocesses with real payloads. Re-running the reviewer's own mutation list after the fixes kills **7 of 7**; the eighth is unreachable by construction and that is stated in the code rather than left looking like covered ground.

This hit the review escalation cap at three rounds and was shipped on an explicit decision to do so. The residual risk is bounded by design: both hooks are advisory-only and always exit 0, and `review_enforcement_commit.py` remains byte-identical to `origin/main`.

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
@coderabbitai review

**Comment by WingedGuardian** [comment]
## Leak / privacy scan — `416a0055`

Scanned the full branch diff against `origin/main`: 1,777 added lines across
4 files (a new script, its test file, and two doc/registration edits). Added
lines only — removals cannot introduce a disclosure.

| Category | Matches |
|---|---|
| IPv4 addresses | 0 |
| IPv6 ULA (`fd00::/8`) | 0 |
| Home paths embedding a username | 0 |
| Credential-shaped strings (api key / secret / bearer / `ghp_` / `sk-`) | 0 |
| Install-specific identifiers (hostnames, account names, remote hosts) | 0 |
| Inferential personal context (household, family, languages, schedule) | 0 |

### Surfaces checked beyond prose

Comments, docstrings and test fixtures were scanned along with the code.
Fixtures are synthetic throughout — the strings under test are invented prose
(`"a shared sentence of prose"`, `"sentence number N of removed prose"`) and the
repositories are `tmp_path` scratch repos, never a copy of this install's tree.

### Measurements quoted in the code and commit message

The scanner's comments cite figures derived from this repository's own public
commit history (95 class-divergences over 1,584 commits; a per-literal timing
curve; recall and fire-rate counts). These describe the repo and the software's
behaviour on it. Two source files are named as examples where a construct
occurs — both are tracked files already public in this repository.

### Paths

The diff references repo-relative paths only. No absolute path, no `~`-prefixed
path, and no reference to a worktree location outside the repository.

VERDICT: PASS
PII/Secrets/Wording: CLEAN

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
## Scheduled review — leaks / private data

Scanned the full branch diff against `origin/main` (2,234 lines) and, separately,
the added lines in isolation.

**Direct identifiers** — RFC1918 ranges, IPv6 ULA, CGNAT/tailnet, home-directory
paths embedding a username, email addresses, SSH/PEM key material: **0 genuine
matches**. The single email-shaped hit was `@pytest.fixture`, which matches the
pattern and is not an address. No `/home/<user>` path appears in any added line.

**Inferential context** — household composition, location, schedule, first-person
"my/our" framing: **0 matches**.

**Test fixtures** were checked specifically, since that is where this class
usually slips through rather than in prose. Git identities in the new tests are
the synthetic `t@t` / `t`, and every repository, path and literal used is
constructed inside `tmp_path`. No real install data is referenced.

PII/Secrets: CLEAN
Wording: CLEAN
VERDICT: PASS

File: scripts/hooks/class_scope_advisory.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        )
+
+    if lines:
+        print(
+            "[class-scope] This edit may have changed one member of a set:\n"
+            + "\n".join(lines)
+            + "\n  Enumerate the rest before moving on — fixing the named instance "
+            "and leaving siblings is what turns one review finding into three "
+            "rounds.",
+            file=sys.stderr,
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/hooks/class_scope_advisory.py]: Resolved by deletion, not by a patch. Both advisory hooks were removed in `416a0055`; the scanner is now a script run deliberately (`scripts/class_scope_scan.py`), so there is no hook output channel to get wrong and no `PostToolUse` registration in `.claude/settings.json`.

That was the point of the conversion rather than a side effect: both P1s on this PR came from the hook-protocol surface, which a script does not have. The finding was correct against the code it reviewed.

File: scripts/hooks/class_scope_commit_advisory.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    if notes:
+        # The documented PreToolUse channel, NOT stderr: stderr from a hook that
+        # exits 0 is discarded, so the advisory would be inert. permissionDecision
+        # stays "allow" — this never blocks.
+        json.dump(
+            {
+                "hookSpecificOutput": {
+                    "hookEventName": "PreToolUse",
+                    "permissionDecision": "allow",
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/hooks/class_scope_commit_advisory.py]: Correct, and resolved by deleting the file rather than by removing the key. `scripts/hooks/class_scope_commit_advisory.py` no longer exists as of `416a0055`; `grep -rn permissionDecision` over what this PR now ships returns nothing, and `.claude/settings.json` carries no `class_scope` registration.

Worth recording why this one existed: it was introduced by my own fix to the round-1 finding on the sibling file — the same fix written two different ways across two files. That is the defect class this scanner is built to detect, in the scanner’s own review history.

**Comment by WingedGuardian** [comment]
## Scheduled review — leaks / private data (re-scan at new head)

Delta since the previously cleared head is two files, +76/-3: an output-stream
reconfigure in `main()`, a docstring correction, and one subprocess test.

The one notable item for this review is deliberate and safe: the new test writes
a NON-ASCII literal (an accented word) into a scratch repository under
`tmp_path` in order to exercise an ASCII stdout. It is invented prose, carries
no identity, and never leaves the temporary directory.

Re-scanned the full branch diff and the added lines in isolation against the
same pattern set: RFC1918 ranges, IPv6 ULA, CGNAT/tailnet, home-directory paths
embedding a username, email addresses, key material, long token-shaped
literals, and the inferential set (household, location, schedule, devices,
first-person framing). **0 genuine matches.** Git identities in the new test are
the synthetic `t@t` / `t`; every path is constructed inside the fixture.

PII/Secrets: CLEAN
Wording: CLEAN
VERDICT: PASS

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
Closing unmerged. The scanner will ship as two smaller PRs rebuilt on stdlib front-ends; this branch's own history is the evidence for why, including the part that argues against the framing I started with.

## What is open at head

Ten Codex P2s are live at `edbed649`, plus one CodeRabbit minor. Classified by mechanism rather than by site:

| mechanism | findings |
|---|---|
| reports clean for a question it never asked | `:102` a `SyntaxError` becomes zero literals · `:804` deleted files · empty files · a `.py`→non-`.py` rename · `:941` more than five survivors truncated silently — **5** |
| identity / multiplicity collapse | `:143` one core kept per value · `:500` the qualname map is last-one-wins · `:646` `str.splitlines()` — 3 |
| incomplete grammar enumeration | `:638` `_scope_binds` models six binding forms · `:211` the prose classifier — 2 |
| wrong snapshot | `:247` `--staged` reads the index for the file and the working tree for the survivor search — 1 |

Five of these were re-read against source before writing this, rather than taken from the comment titles; all five are real. Two — `:247` and `:646` — are independent of any engine choice and were absent from my own first analysis.

The first row also has two members no reviewer named. Measured with one repository per change shape, the sibling always holding the literal:

| change shape | scanner output | |
|---|---|---|
| modify | `1 of 1 changed file(s) scanned; 1 finding` | control |
| delete | `no modified Python files` | wrong |
| rename `.py`→`.md` | `no modified Python files` | wrong |
| empty at head | `0 of 1 scanned`, labelled unreadable | wrong |

A deleted file and a rename out of Python both produce an affirmatively false clean statement.

## The convergence argument was tested, and it did not hold

My first diagnosis was that hand-rolled Python semantics acts as a generator that prevents convergence. That is not what this repository's history shows:

| PR | what it is | added | rounds | findings per round | outcome |
|---|---|---|---|---|---|
| #1537 | novel hand-rolled `ast` static analyser | 1,129 | 3 | 7 → 6 → 6 | merged |
| #1527 | guard | 1,187 | 4 | 6 → 1 → 1 → 2 | merged |
| this PR | novel analyser | 2,273 | 4 | 12 → 7 → 9 → 1 | closing |

#1537 hand-rolls the same class of scope analysis, drew a higher finding density (16.8 per thousand added lines against this PR's 12.8), never converged in count, and merged. Four rounds is the ordinary cost of a novel static analyser here — not evidence of a broken design — and the replacement should not be expected to converge faster.

This PR is closed for the defects above and for two foundations that measure better, not because it could not have been finished.

## What replaces it

Both scans ask semantic questions and answer them with syntax. CPython's own front-ends answer them directly. Measured across 1,393 files in this repository:

**Literal spellings — `ast` node spans intersected with `tokenize` token positions.** 99.79% of 90,453 string-constant nodes recover their exact source spelling; the 188 that do not become an explicit skip rather than a silent one. 2,845 nodes carry more than one spelling — the population `:143` names, which the present `source_cores` cannot see because it keeps a single longest core per value. The intersection also excludes trailing comments: `ast.get_source_segment` returns a span, so an implicit concatenation with an interleaved `# noqa` currently yields a core that greps the comment. A `tokenize`-only map was tried first and rejected — it loses 15.95% of literals, because implicit concatenation joins into a value that is never a single token.

**Binding and scope — `symtable`.** 1,393 of 1,393 files parse. It models twelve binding forms against `_scope_binds`'s six, including `import`, `import as`, `with as`, `except as`, `match` captures and star-patterns — the forms that have been arriving one per round. It states the two deliberately opposite scope policies natively instead of re-deriving them, and its children are a list, so both members of the `@property` / `@x.setter` pair `:500` describes survive. Scope-to-node joins are ambiguous in 1 of 10,961 nested scopes; that case resolves conservatively, as a skip.

## Retained

The measurement corpora and harnesses, and the numbers they produced: 95 literal-side positives mined from 1,584 commits, 74 provenance-side positives, and the recall and fire-rate baselines. These are implementation-independent and become the acceptance bar for the replacement — a rebuild that improves correctness while losing recall does not ship.

Also retained: the git and CLI layer, which is orthogonal to both engines and has already been through four rounds; the reporting invariant that every literal ends in exactly one of finding, skipped-with-reason, or verified-clean; and the methodology documentation, which merged separately in #1532.

No changes from this branch are being merged. The branch is left in place.
